### PR TITLE
Sync ``libasr`` with LFortran

### PR DESCRIPTION
**Pull request by czgdp1807**
Corresponding LFortran PR - https://github.com/lfortran/lfortran/pull/1681

### CODE REVIEW

File: src/libasr/codegen/asr_to_c_cpp.h [hunk 1]
```diff
@@ -451,9 +443,6 @@ R"(#include <stdio.h>
             } else if (ASR::is_a<ASR::TypeParameter_t>(*return_var->m_type)) {
                 has_typevar = true;
                 return "";
-            } else if (ASR::is_a<ASR::Dict_t>(*return_var->m_type)) {
```
Comment (Smit-create) [inline src/libasr/codegen/asr_to_c_cpp.h:454]: This needs to be reverted.

File: src/libasr/codegen/asr_to_c_cpp.h [hunk 2]
```diff
@@ -550,16 +539,9 @@ R"(#include <stdio.h>
             src = "";
             return;
         }
-        ASR::FunctionType_t *f_type = ASRUtils::get_FunctionType(x);
```
Comment (Smit-create) [inline src/libasr/codegen/asr_to_c_cpp.h:553]: This needs to be reverted.

File: src/libasr/codegen/asr_to_c_cpp.h [hunk 3]
```diff
@@ -1390,18 +1349,12 @@ R"(#include <stdio.h>
                 // src = src;
                 break;
             }
-            case (ASR::cast_kindType::IntegerToUnsignedInteger) : {
```
Comment (Smit-create) [inline src/libasr/codegen/asr_to_c_cpp.h:1393]: This needs to be reverted.

File: src/libasr/codegen/asr_to_c_cpp.h [hunk 4]
```diff
@@ -1536,10 +1489,6 @@ R"(#include <stdio.h>
         handle_Compare(x);
     }
 
-    void visit_UnsignedIntegerCompare(const ASR::UnsignedIntegerCompare_t &x) {
```
Comment (Smit-create) [inline src/libasr/codegen/asr_to_c_cpp.h:1539]: This needs to be reverted.

File: src/libasr/codegen/asr_to_c_cpp.h [hunk 5]
```diff
@@ -1811,56 +1748,23 @@ R"(#include <stdio.h>
 
     void visit_Allocate(const ASR::Allocate_t &x) {
```
Comment (Smit-create) [inline src/libasr/codegen/asr_to_c_cpp.h:1749]: Everything inside `visit_Allocate` needs to be reverted.

File: src/libasr/codegen/c_utils.h [hunk 6]
```diff
@@ -1175,6 +1175,9 @@ class CCPPDSUtils {
         }
 
         std::string get_dict_type(ASR::Dict_t* dict_type) {
+            if (!ASR::is_a<ASR::Integer_t>(*dict_type->m_key_type)) {
```
Comment (Smit-create) [inline src/libasr/codegen/c_utils.h:1178]: This needs to be reverted.

File: src/libasr/codegen/asr_to_llvm.cpp [hunk 7]
```diff
@@ -3513,38 +3573,10 @@ class ASRToLLVMVisitor : public ASR::BaseVisitor<ASRToLLVMVisitor>
                 }
                 break;
             }
-            case (ASR::ttypeType::UnsignedInteger) : {
```
Comment (Thirumalai-Shaktivel) [inline src/libasr/codegen/asr_to_llvm.cpp:3516]: This needs to be reverted: https://github.com/lcompilers/lpython/pull/1764/files#diff-48a7161397f9eb8b730f0cbec2f11ce1471aac961dc6b16dc558cf161c375054

File: src/libasr/codegen/asr_to_llvm.cpp [hunk 8]
```diff
@@ -4468,60 +4538,171 @@ class ASRToLLVMVisitor : public ASR::BaseVisitor<ASRToLLVMVisitor>
...
+                    llvm::Type* llvm_data_type = get_type_from_ttype_t_util(data_type);
+                    fill_array_details(llvm_target_, llvm_data_type, m_dims, n_dims, false, false);
+                    builder->CreateStore(llvm_value, arr_descr->get_pointer_to_data(llvm_target_));
+                    llvm_value = llvm_target_;
+                }
+            builder->CreateStore(llvm_value, llvm_target);
+            builder->CreateStore(llvm_value, llvm_target);
             builder->CreateStore(llvm_value, llvm_target);
+            builder->CreateStore(llvm_value, llvm_target);
+            builder->CreateStore(llvm_value, llvm_target);
```
Comment (Thirumalai-Shaktivel) [inline src/libasr/codegen/asr_to_llvm.cpp:4689]: I think the repeated lines are to be reverted.

File: src/libasr/pass/array_op.cpp [hunk 9]
```diff
@@ -310,11 +327,6 @@ class ReplaceArrayOp: public ASR::BaseExprReplacer<ReplaceArrayOp> {
                             al, loc, left, (ASR::binopType)x->m_op,
                             right, x->m_type, nullptr));
 
-            case ASR::exprType::UnsignedIntegerBinOp:
```
Comment (Thirumalai-Shaktivel) [inline src/libasr/pass/array_op.cpp:313]: To be reverted: https://github.com/lcompilers/lpython/pull/1775/files#diff-51616d9e7a321bfa6245084060933b377e1ab9e391b2b4f248fd9d8ee9bf2ce0

File: src/libasr/pass/array_op.cpp [hunk 10]
```diff
@@ -674,10 +856,6 @@ class ReplaceArrayOp: public ASR::BaseExprReplacer<ReplaceArrayOp> {
         replace_ArrayOpCommon<ASR::IntegerCompare_t>(x, "_integer_comp_op_res");
     }
 
-    void replace_UnsignedIntegerCompare(ASR::UnsignedIntegerCompare_t* x) {
```
Comment (Thirumalai-Shaktivel) [inline src/libasr/pass/array_op.cpp:677]: To be reverted.

File: src/libasr/pass/intrinsic_function_registry.h [hunk 11]
```diff
@@ -727,9 +810,13 @@ namespace Abs {
             ASR::symbol_t *s = scope->get_symbol(func_name);
             ASR::Function_t *f = ASR::down_cast<ASR::Function_t>(s);
             return b.Call(s, new_args, expr_type(f->m_return_var),
-                compile_time_value);
+                                compile_time_value);
         }
+
+        func_name = scope->get_unique_name(func_name);
+
```
Comment (Thirumalai-Shaktivel) [inline src/libasr/pass/intrinsic_function_registry.h:817]:
```suggestion
```
get_unique_name is present in the `declare_basic_variables` macro.

Comment (Thirumalai-Shaktivel) [reply src/libasr/pass/intrinsic_function_registry.h:817]: I can help you with the `intrinsic_function_registry` and I also apologize for not submitting the sync PR with these changes to LFortran, I will submit it soon.

Comment (czgdp1807) [reply src/libasr/pass/intrinsic_function_registry.h:817]: On my system everything worked with LPython after the sync. I am yet to test it with LFortran.

File: src/lpython/parser/semantics.h [hunk 12]
```diff
@@ -798,8 +798,8 @@ static inline ast_t* concat_string(Allocator &al, Location &l,
         x.c_str(p.m_a), expr_contextType::Load)
 // `x.int_n` is of type BigInt but we store the int64_t directly in AST
 #define INTEGER(x, l) make_ConstantInt_t(p.m_a, l, x, nullptr)
-#define STRING1(x, l) make_ConstantStr_t(p.m_a, l, unescape_string(p.m_a, x), nullptr)
-#define STRING2(x, y, l) concat_string(p.m_a, l, EXPR(x), unescape_string(p.m_a, y), nullptr)
+#define STRING1(x, l) make_ConstantStr_t(p.m_a, l, LCompilers::s2c(p.m_a, LCompilers::unescape_string(p.m_a, x.str())), nullptr)
```
Comment (Thirumalai-Shaktivel) [inline src/lpython/parser/semantics.h:801]: The changes in this PR need to be updated here, I think. https://github.com/lcompilers/lpython/pull/1762/files#diff-5a134f23f49700449fdd6d749a06b72ca2c844ed6c30fa752d6194b4b8b5efb7

File: src/libasr/string_utils.cpp [hunk 13]
```diff
@@ -154,20 +154,15 @@ std::string join_paths(const std::vector<std::string> &paths) {
     return p;
 }
 
-char* unescape_string(Allocator &al, LCompilers::Str &s) {
```
Comment (Thirumalai-Shaktivel) [inline src/libasr/string_utils.cpp:157]: ..

File: src/lpython/parser/semantics.h (hunk 12 above)
Comment (czgdp1807) [reply src/lpython/parser/semantics.h:801]: Could you please explain? I think the only change here is in the API of `unescape_string`. It used to return `char*` in LPython but when synced with LFortran it returns `std::string`, so its usage has to be updated everywhere in LPython.

File: src/libasr/codegen/asr_to_llvm.cpp [hunk 14]
```diff
@@ -6180,7 +6421,8 @@ class ASRToLLVMVisitor : public ASR::BaseVisitor<ASRToLLVMVisitor>
     }
 
     void visit_StringConstant(const ASR::StringConstant_t &x) {
-        tmp = builder->CreateGlobalStringPtr(x.m_s);
+        std::string s = unescape_string(al, x.m_s);
```
Comment (Thirumalai-Shaktivel) [inline src/libasr/codegen/asr_to_llvm.cpp:6424]: unescape_string is handled in the frontend itself: https://github.com/lcompilers/lpython/pull/1762/files#diff-48a7161397f9eb8b730f0cbec2f11ce1471aac961dc6b16dc558cf161c375054

File: src/libasr/string_utils.cpp (hunk 13 above)
Comment (Thirumalai-Shaktivel) [reply src/libasr/string_utils.cpp:157]: This function seems to be updated.

File: src/lpython/parser/semantics.h (hunk 12 above)
Comment (Thirumalai-Shaktivel) [reply src/lpython/parser/semantics.h:801]: LFortran uses this function in asr_to_llvm, I think that needs to be reverted.
So, the changes related to unescape can be reverted.

**Comment by Thirumalai-Shaktivel** [comment]
@czgdp1807 Should I submit a Sync PR to LFortran for the `IntrinisicFunction` changes or you will submit it through "Sync libasr with LPython" PR?

File: src/lpython/parser/semantics.h (hunk 12 above)
Comment (Thirumalai-Shaktivel) [reply src/lpython/parser/semantics.h:801]: See https://github.com/lcompilers/lpython/pull/1502#issuecomment-1535080906

**Comment by czgdp1807** [comment]
Let us complete the sync process first before making any updates elsewherw.

File: src/libasr/codegen/wasm_to_x64.cpp [hunk 15]
```diff
@@ -582,6 +582,8 @@ class X64Visitor : public WASMDecoder<X64Visitor>,
     void visit_F32Sqrt() { visit_F64Sqrt(); }
 
     void gen_x64_bytes() {
+        emit_elf64_header(m_a);
+
```
Comment (Shaikh-Ubaid) [inline src/libasr/codegen/wasm_to_x64.cpp:586]: All the changes in the files `wasm_to_x64.cpp`, `x86_assembler.cpp` and `x86_assembler.h` needs to be reverted.

**Comment by czgdp1807** [comment]
I think Windows test fail due to changes related to `unescape_string`. @Thirumalai-Shaktivel Can you please check?

File: src/libasr/asdl_cpp.py [hunk 16]
```diff
@@ -1615,12 +1657,12 @@ def visitField(self, field, cons):
...
                 if field.opt:
                     self.emit("if (x.m_%s) {" % field.name, 2)
-                    self.emit(    's.append("\\"" + get_escaped_str(x.m_%s) + "\\"");' % field.name, 3)
+                    self.emit(    's.append("\\"" + std::string(x.m_%s) + "\\"");' % field.name, 3)
                     self.emit("} else {", 2)
                     self.emit(    's.append("()");', 3)
                     self.emit("}", 2)
                 else:
-                    self.emit('s.append("\\"" + get_escaped_str(x.m_%s) + "\\"");' % field.name, 2)
+                    self.emit('s.append("\\"" + std::string(x.m_%s) + "\\"");' % field.name, 2)
```
Comment (Shaikh-Ubaid) [inline src/libasr/asdl_cpp.py:1665]: I think the issue is here. This needs to be reverted.

**Review by Thirumalai-Shaktivel — APPROVED** [review]

File: src/libasr/codegen/asr_to_llvm.cpp [hunk 17]
```diff
@@ -4468,60 +4566,155 @@ class ASRToLLVMVisitor : public ASR::BaseVisitor<ASRToLLVMVisitor>
...
+                    ASR::ttype_t* data_type = ASRUtils::duplicate_type_without_dims(
+                                                al, target_type, target_type->base.loc);
+                    llvm::Type* llvm_data_type = get_type_from_ttype_t_util(data_type);
+                    fill_array_details(llvm_target_, llvm_data_type, m_dims, n_dims, false, false);
+                    builder->CreateStore(llvm_value, arr_descr->get_pointer_to_data(llvm_target_));
+                    llvm_value = llvm_target_;
+                }
+            builder->CreateStore(llvm_value, llvm_target);
             builder->CreateStore(llvm_value, llvm_target);
+                builder->CreateStore(llvm_value, llvm_target);
```
[the remark this answers is not in the document]
Comment (czgdp1807) [reply src/libasr/codegen/asr_to_llvm.cpp:4715]: This is very weird. I removed the repetition but still its showing up every time I push.

**Review by Smit-create — APPROVED** [review]
Thanks!!

File: src/libasr/pass/pass_manager.h [hunk 18]
```diff
@@ -212,8 +216,8 @@ namespace LCompilers {
             _with_optimization_passes = {
                 "global_stmts",
                 "init_expr",
-                "implied_do_loops",
                 "class_constructor",
+                "implied_do_loops",
```
Comment (Shaikh-Ubaid) [inline src/libasr/pass/pass_manager.h:220]: This needs to be reverted. Sorry, I just noticed it while reviewing the sync PR at LFortran.

**Review by Pranavchiku — APPROVED** [review]
LGTM! Thank you for the sync :)

**Review by Shaikh-Ubaid — APPROVED** [review]
It looks good to me. Thank you so much for this! Also, please see https://github.com/lfortran/lfortran/pull/1681#discussion_r1200888168.

**Pull request merged**